Add oracle

TheRealFalcon opened the pull request:
Also includes an additional commit for some small key handling refactors that I don't want squashed. Might make sense to review the two commits separately.

The API purposely doesn't completely line up with the base class (like many of the other implementing classes) but is taking the form that I would like the base class to use (in a separate soon-to-come PR). It should work as-is though.

TheRealFalcon commented:
tox failure is mostly me forgetting how to format docstrings. I'll fix that up soon but I don't think it should block review.

blackboxsw commented on pycloudlib/oracle/cloud.py:
```diff
@@ -0,0 +1,218 @@
+        super(OCI, self).__init__(tag)
+        self.compartment_id = compartment_id
+        self.config_path = config_path
+        config = oci.config.from_file(str(config_path))
```
Can we do a pre-flight if not os.exists(config_path) here and raise a RuntimeError saying "get your OCI client setup: see <LAUNCHING_ORACLE_URL>?

blackboxsw commented on pycloudlib/oracle/cloud.py:
```diff
@@ -0,0 +1,218 @@
+        """
+        self.compute_client.delete_image(image_id)
+
+    def released_image(self, release, operating_system='Canonical Ubuntu'):
```
Do you have current use-cases in mind to have call-sites currently where we are supporting non-canonical Ubuntu?
Are you planning for centOS integration testing for cloud-init specifically?

TheRealFalcon replied:
No, I initially hardcoded things to be Ubuntu, but thought for a library that might be used outside of just the cloud-init project, we might need greater flexibility. Do you have a concern with doing it this way, or just curious?

blackboxsw replied:
This was only me being curious, thanks. I wasn't sure if you already has another specific consumer in mind.

blackboxsw commented on pycloudlib/util.py:
```diff
@@ -280,3 +280,13 @@ def _safe_int(possible_int):
         return None
+
+
+def ubuntu_name_to_number(name):
+    """Convert to version number based on the version name."""
```
Any objection to a global dict? The function seems a bit over the top of a dict lookup. /me was tempted to suggest distro-info, but that's an external dep dependency. 

UBUNTU_RELEASE_VERSION_MAP = {
    'focal': '20.04',
    'bionic': '18.04',
    'xenial': '16.04',
}

TheRealFalcon replied:
Not familiar with distro-info. I'm willing to entertain an external dependency, as that can be automatically updated. I'm not really a big fan of what I did here because it means it will need to be updated after every new release. Global dict makes sense though.

blackboxsw replied:
No need for this. If it becomes a maintenance burden we can look at using an external dynamic utility for this, but python-proper libraries or modules are probably limited here I think as most modules would want to know "what environment I'm running in" instead of "what is the known set/map of all releases available".

blackboxsw commented on pycloudlib/oracle/utils.py:
```diff
@@ -0,0 +1,25 @@
+            return current_data
+        time.sleep(1)
+    raise Exception(
+        "Still not ready. "
```
Static exception messages give me the grumpies. Can we parameterize the message a bit to say desired state <state> not met. Still <current_state> after waiting <sleep_seconds>"

TheRealFalcon commented:
@blackboxsw I believe I've addressed all of your comments and there are 2 open questions.

blackboxsw commented on docs/clouds/oci.md:
```diff
@@ -0,0 +1,63 @@
+
+##  Instances
+
+Lauching instances requires at minimum an image_id, though instance_type
```
```suggestion
Launching instances requires at minimum an image_id, though instance_type
```

blackboxsw commented on pycloudlib/oci/cloud.py:
```diff
@@ -0,0 +1,227 @@
+            string, id of latest image
+
+        """
+        if operating_system == 'Canonical Ubuntu':
```
Not a blocker to release of this PR, What in your mind differentiates 'Canonical Ubuntu' from just 'ubuntu'? I'm thinking about things in terms of when your followup refactor-api tries to generalize methods up in pycloudlib.cloud parent class, and ease of use when folks try to run 'ubuntu' vs. 'opensuse' 'debian' or others. 

I realize this might just be an Oracle CLI need, for which I'd say we'd probably like to standardize on something simple in the parent class, that get's translated to "Canonical Ubuntu" in the oracle api call

TheRealFalcon replied:
This is just a consequence of the Oracle API. If I want to list Ubuntu images with the `self.compute_client.list_images(...)` call, I have to pass 'Canonical Ubuntu' as the `operating_system` argument. In my head, this was an Oracle-specific thing. In my refactor-api branch, I was expecting the method to look like `def daily_image(self, release, **kwargs)`. Since the default is defined here, that's how the API will normally function, but if you have a reason to make some Oracle-specific calls, you can override which operating_system selection.

TheRealFalcon commented:
@blackboxsw I fixed the typo, rebased and squashed (I don't want any additional squashing). Since I made more changes I need another +1 before it'll let me merge.
